common/test/TestLogging.cs

```diff
@@ -2,39 +2,115 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
-using System.Diagnostics;
-using System.Globalization;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Versioning;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Microsoft.Azure.Devices.E2ETests
 {
+    /// <summary>
+    /// This class logs to multiple providers - Event source, Application insights
+    /// </summary>
     public class TestLogging
     {
-        private readonly static TestLogging s_instance = new TestLogging();
+        public const string Language = "C#";
+        public const string Service = "IotHub";
+        private readonly object initLock = new object();
+
+        // Client to log to application insights
+        private static TelemetryClient _telemetryClient;
+
+        // Unique ID for all tests of a run.
+        private static string _testRunId;
+
+        public TestContext Context { get; set; }
+        public string TargetFramework { get; set; }
 
         private TestLogging()
         {
+            // Thread-safe initialization of the telemetry client and run Id.
+            lock (initLock)
+            {
+                if (_telemetryClient == null)
+                {
+                    string iKey = Environment.GetEnvironmentVariable("E2E_IKEY");
+                    if (!string.IsNullOrWhiteSpace(iKey))
+                    {
+                        var config = new TelemetryConfiguration
+                        {
+                            InstrumentationKey = iKey,
+                        };
+                        _telemetryClient = new TelemetryClient(config);
+                        _testRunId = Guid.NewGuid().ToString();
+                    }
+                }
+            }
         }
 
         private const string NullInstance = "(null)";
-        
+
         public static string IdOf(object value) => value != null ? value.GetType().Name + "#" + GetHashCode(value) : NullInstance;
 
         public static int GetHashCode(object value) => value?.GetHashCode() ?? 0;
 
-        public static TestLogging GetInstance()
+        public static TestLogging GetInstance(TestContext testContext = null)
```

Comment:
This ties us to a specific testing framework ,doesn't it? Perhaps break out the parameters you specifically want.

Reply:
It does tie us to a framework. I did have it like that previously and it became more than one parameter and decided to pass the whole object. But I do agree it is probably better not to tie it to a specific framework. Will send required parameters.

Reply:
We can always change it later, I guess.